distill/process/log.py

```diff
@@ -0,0 +1,83 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+from pydantic import BaseModel, Field, ValidationError
+from pydantic.config import ConfigDict
+
+from decimal import Decimal
+from typing import List, Union, Any
+from enum import Enum
+
+class Browser(BaseModel):
+    browser: str
+    version: str
+
+class Location(BaseModel):
+    x: Union[int, None]
+    y: Union[int, None]
+
+class ScrnRes(BaseModel):
+    width: int
+    height: int 
+
+class LogType(str, Enum):
+    raw = "raw"
+    custom = "custom"
+
+class Details(BaseModel):
+    window: bool
+
+class LogSchema(BaseModel):
+    """
+    A raw or custom log produced by userale
+    """
+    model_config = ConfigDict(title="Log")
+
+    target: str
+    path: List[str]
+    pageUrl: str 
+    pageTitle: str
+    pageReferrer: str
+    browser: Browser
+    clientTime: int 
+    microTime: int = Field(..., lt=2)
+    location: Location 
+    scrnRes: ScrnRes 
+    type_field: str = Field(..., alias="type", serialization_alias="type")
+    logType: LogType
+    userAction: bool
+    details: Details
+    userId: str 
+    toolVersion: Union[str, None] 
+    toolName: Union[str, None]
+    useraleVersion: Union[str, None] 
+    sessionID: str 
+
+class Log():
+    def __init__(self, log_data: str):
```

Comment:
The idea is that the log should accept a schema as an argument which will then be used to validate the incoming data. This should default to our UserALE schema, but users should be able to pass their own.

Also, not all incoming data will necessarily be a `str`. If your implicit assumption is that `log_data` will be a json serialized str, then you should make that assumption explicit in the argument name, i.e. `json_data` or `json`.

We should be able to handle dicts too. For instance, the following should work:
```py
plain_dict_log = userale_log.deserializeJson() # assume this is a Log instance already
userale_log_2 = Log(raw_data=plain_dict_log)
```
You could use a separate, optional argument for this, `raw_data` or you could create a single, overloaded `data` argument and then conditionally handle inputs within the init logic. We can discuss pros/cons of this.

Reply:
data should be either a json parsable string or a dict (also parsable), conditionally check type of argument coming in to parse correctly. Can raise error based on bad types

Reply:
make schema argument optional and default to userale schema

Reply:
also add docstring